File: solver/rubiks-cube-pattern-key.cpp
#include "rubiks-cube-pattern-key.h"

// Constructor
RubiksCubePatternKey::RubiksCubePatternKey(RubiksCubeInformation* rubiks_cube_information) {
    this->information = rubiks_cube_information;
}

// Create 40-bit key from corner positions and orientations
uint64_t RubiksCubePatternKey::fromRubiksCubeToCornersKey(RubiksCube& rubiks_cube) {
    uint64_t corner_key = 0;
    for (int corner_position = 0; corner_position < 8; ++corner_position) {
        int corner_type = rubiks_cube.getCornerType(corner_position);
        int orientation = rubiks_cube.getCornerOrientation(corner_position);
        corner_key <<= 5;
        corner_key |= ((orientation & 0b11) << 3) | (corner_type & 0b111);
    }
    return corner_key;
}

// Create 60-bit key from edge positions and orientations
uint64_t RubiksCubePatternKey::fromRubiksCubeToAllEdgesKey(RubiksCube& rubiks_cube) {
    uint64_t all_edges_key = 0;
    for (int edge_position = 0; edge_position < 12; ++edge_position) {
        int edge_type = rubiks_cube.getEdgeType(edge_position);
        int orientation = rubiks_cube.getEdgeOrientation(edge_position);
        all_edges_key <<= 5;
        all_edges_key |= ((orientation & 0b1) << 4) | (edge_type & 0b1111);
    }
    return all_edges_key;
}

// Load corner types and orientations from 40-bit key
RubiksCube RubiksCubePatternKey::fromCornersKeyToRubiksCube(uint64_t corner_key) {
    RubiksCube rubiks_cube(this->information, false);
    for (int i = 0; i < 40; i += 5) {
        uint64_t chunk = (corner_key >> (35 - i)) & 0b11111;
        int corner_position = i / 5;
        int corner_type = chunk & 0b111;
        int orientation = (chunk & 0b11000) >> 3;
        rubiks_cube.setCornerType(corner_position, corner_type, orientation);
    }
    return rubiks_cube;
}

// Load edge types and orientations from 60-bit key
RubiksCube RubiksCubePatternKey::fromAllEdgesKeyToRubiksCube(uint64_t edge_key) {
    RubiksCube rubiks_cube(this->information, false);
    for (int i = 0; i < 60; i += 5) {
        uint64_t chunk = (edge_key >> (55 - i)) & 0b11111;
        int edge_position = i / 5;
        int edge_type = chunk & 0b1111;
        int orientation = (chunk & 0b10000) >> 4;
        rubiks_cube.setEdgeType(edge_position, edge_type, orientation);
    }
    return rubiks_cube;
}

// Convert all edges key to edges key for edges with type <= 5
uint64_t RubiksCubePatternKey::fromAllEdgesKeyToEdgesKey(uint64_t all_edges_key) {
    uint64_t edges_key = 0;
    for (int i = 0; i < 60; i += 5) {
        const uint64_t chunk = (all_edges_key >> (55 - i)) & 0b11111;
        const int edge_type = chunk & 0b1111;
        edges_key <<= 4;
        if (edge_type <= 5) {
            const int orientation = (chunk & 0b10000) >> 4;
            edges_key |= ((orientation & 0b1) << 3) | (edge_type & 0b111);
        } else {
            edges_key |= 6;
        }
    }
    return edges_key;
}

// Convert all edges key to edges key for edges with type >= 6
uint64_t RubiksCubePatternKey::fromAllEdgesKeyToOtherEdgesKey(uint64_t all_edges_key) {
    uint64_t edges_key = 0;
    for (int i = 0; i < 60; i += 5) {
        uint64_t chunk = (all_edges_key >> (55 - i)) & 0b11111;
        int edge_type = chunk & 0b1111;
        edges_key <<= 4;
        if (edge_type >= 6) {
            int orientation = (chunk & 0b10000) >> 4;
            edges_key |= ((orientation & 0b1) << 3) | ((edge_type - 6) & 0b111);
        } else {
            edges_key |= 6;
        }
    }
    return edges_key;
}

RubiksCube * RubiksCubePatternKey::fromKeysToRubiksCube(uint64_t cornerKey, uint64_t edgeKey, uint64_t otherEdgeKey) {
    auto * cube = new RubiksCube(this->information, false);
    for (int i = 0; i < 40; i += 5) {
        const uint64_t chunk = (cornerKey >> (35 - i)) & 0b11111;
        const int corner_position = i / 5;
        const int corner_type = chunk & 0b111;
        const int orientation = (chunk & 0b11000) >> 3;
        cube->setCornerType(corner_position, corner_type, orientation);
    }
    for (int i = 0; i < 48; i += 4) {
        const uint64_t chunk = (edgeKey >> (44 - i)) & 0b1111;
        const int edge_position = i / 4;
        const int edge_type = chunk & 0b111;
        const int orientation = (chunk & 0b1000) >> 3;
        if (edge_type <= 5) {
            cube->setEdgeType(edge_position, edge_type, orientation);
        }
    }
    for (int i = 0; i < 48; i += 4) {
        uint64_t chunk = (otherEdgeKey >> (44 - i)) & 0b1111;
        int edge_position = i / 4;
        int edge_type = chunk & 0b111;
        int orientation = (chunk & 0b1000) >> 3;
        if (edge_type <= 5) {
            cube->setEdgeType(edge_position, edge_type + 6, orientation);
        }
    }
    return cube;
}

array<uint64_t, 3> RubiksCubePatternKey::fromRubiksCubeToKeys(RubiksCube & rubiksCube) {
    const uint64_t allEdgesKey = fromRubiksCubeToAllEdgesKey(rubiksCube);
    return {
        fromRubiksCubeToCornersKey(rubiksCube),
        fromAllEdgesKeyToEdgesKey(allEdgesKey),
        fromAllEdgesKeyToOtherEdgesKey(allEdgesKey)};
}

// Print corner key in binary, 5 bits at a time
void RubiksCubePatternKey::printCornerKey(uint64_t key) {
    for (int i = 0; i < 40; i += 5) {
        uint64_t chunk = (key >> (35 - i)) & 0b11111;
        std::bitset<5> bits(chunk);
        std::cout << bits.to_string() << std::endl;
    }
}

File: heuristics-generator/rubiks_cube_pattern_key.py
from rubiks_cube import RubiksCube

class RubiksCubePatternKey:
    def __init__(self, rubiks_cube_information):
        self.information = rubiks_cube_information

    # create 40 bit key from corner positions and orientations 
    # each 5 bits in the key correspond to corner type (3 bits) and orientation (2 bits)
    def from_rubiks_cube_to_corners_key(self, rubiks_cube: RubiksCube):
        corner_key = 0
        for corner_position in range(8):
            corner_type = rubiks_cube.get_corner_type(corner_position)
            orientation = rubiks_cube.get_corner_orientation(corner_position)
            corner_key <<= 5
            corner_key |= ((orientation & 0b11) << 3) | (corner_type & 0b111)
        return corner_key

    # create 60 bit key from edge positions and orientations 
    # each 5 bits in the key correspond to edge type (4 bits) and orientation (1 bit)
    def from_rubiks_cube_to_all_edges_key(self, rubiks_cube: RubiksCube):
        all_edges_key = 0
        for edge_position in range(12):
            edge_type = rubiks_cube.get_edge_type(edge_position)
            orientation = rubiks_cube.get_edge_orientation(edge_position)
            all_edges_key <<= 5
            all_edges_key |= ((orientation & 0b1) << 4) | (edge_type & 0b1111)
        return all_edges_key

    # load corner type positions and orientations from 40 bit key
    def from_corners_key_to_rubiks_cube(self, corner_key):
        rubiks_cube = RubiksCube(self.information, False)
        for i in range(0, 40, 5):
            chunk = (corner_key >> (35 - i)) & 0b11111
            corner_position = int(i / 5)
            corner_type = chunk & 0b111
            orientation = (chunk & 0b11000) >> 3
            rubiks_cube.set_corner_type(corner_position, corner_type, orientation)
        return rubiks_cube

    # load edge type positions and orientations from 60 bit key
    def from_all_edges_key_to_rubiks_cube(self, edge_key):
        rubiks_cube = RubiksCube(self.information, False)
        for i in range(0, 60, 5):
            chunk = (edge_key >> (55 - i)) & 0b11111
            edge_position = int(i / 5)
            edge_type = chunk & 0b1111
            orientation = (chunk & 0b10000) >> 4
            rubiks_cube.set_edge_type(edge_position, edge_type, orientation)
        return rubiks_cube

    # create 60 bit key from edge positions and orientations 
    # each 5 bits in the key correspond to edge type (4 bits) and orientation (1 bit)
    def from_all_edges_key_to_edges_key(self, all_edges_key):
        edges_key = 0
        for i in range(0, 60, 5):
            chunk = (all_edges_key >> (55 - i)) & 0b11111
            edge_type = chunk & 0b1111
            edges_key <<= 4
            if edge_type <= 5:
                orientation = (chunk & 0b10000) >> 4
                edges_key |= ((orientation & 0b1) << 3) | (edge_type & 0b111)
            else:
                edges_key |= 6
        return edges_key
    
    # create 60 bit key from edge positions and orientations 
    # each 5 bits in the key correspond to edge type (4 bits) and orientation (1 bit)
    def from_all_edges_key_to_other_edges_key(self, all_edges_key):
        edges_key = 0
        for i in range(0, 60, 5):
            chunk = (all_edges_key >> (55 - i)) & 0b11111
            edge_type = chunk & 0b1111
            edges_key <<= 4
            if edge_type >= 6:
                orientation = (chunk & 0b10000) >> 4
                edges_key |= ((orientation & 0b1) << 3) | ((edge_type - 6) & 0b111)
            else:
                edges_key |= 6
        return edges_key

    def print_corner_key(self, key):
        for i in range(0, 40, 5):
            # extract 5 bits using bitwise shift and mask
            chunk = (key >> (35 - i)) & 0b11111
            # convert the 5-bit chunk to a binary string and print
            print(f"{chunk:05b}")
